### PR TITLE
Add timestamp support in signature metadata

### PR DESCRIPTION
**Pull request by kchalkias**

### CODE REVIEW

**Review by mikehearn — CHANGES_REQUESTED** [review]
Can we get a unit test just to make sure the property is pinned down by something?

Also, `@param` -> `@property`

**Comment by kchalkias** [comment]
This failing test is probably wrong. We shouldn't test if two signatures are equal. This might be true for deterministic EdDSA, but not for the rest of the algorithms, each signature, even for the same inputs might differ, due to random nonces used in signing.

**Comment by kchalkias** [comment]
Hm.. this is a double issue 1) is the deterministic signature b) the signed data has changed due to the timestamp change.

File: core/src/main/kotlin/net/corda/core/crypto/SignatureMetadata.kt [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
  * SignatureMeta is required to add extra meta-data to a transaction's signature.
  * It currently supports platformVersion only, but it can be extended to support a universal digital
  * signature model enabling partial signatures and attaching extra information, such as a user's timestamp or other
  * application-specific fields.
  *
- * @param platformVersion current DLT version.
- * @param schemeNumberID number id of the signature scheme used based on signer's key-pair, see [SignatureScheme.schemeNumberID].
+ * @property platformVersion current DLT version.
+ * @property schemeNumberID number id of the signature scheme used based on signer's key-pair, see [SignatureScheme.schemeNumberID].
+ * @param timestamp an [Instant] that represents the time the signature took place. If not provided, then the current UTC time [Instant.now] is used.
```
Comment (gendal) [inline core/src/main/kotlin/net/corda/core/crypto/SignatureMetadata.kt]: Is this the time the signer _claims_ the signature took place?

Either way, this is great to see added... I fully expect us to be asked to have notaries (single-node or clusters with a leader I guess) state what time their signature should be considered to have been applied at.

Comment (kchalkias) [reply core/src/main/kotlin/net/corda/core/crypto/SignatureMetadata.kt]: great, I will add the clarification that the signer *claims*

**Comment by kchalkias** [comment]
Although this version attaches timestamps in signatures as expected, we don't by any means validate them yet.
There is also a corner case we should consider in a future PR and when we define a timestamping policy.
_"It might be possible that the notary validates a `TimeWindow` before applying its signature, but then the signature's timestamp is slightly outside the range of this window."_ We should always keep a `timestamp` of any received message (always do this first) and then use this `timestamp` on any time-related checks required (like `TimeWindowChecker` and signature `timestamp`).

But again this might not be a good practice and requires a careful consideration per case in Multi-threaded queued tx scenarios, because the clients might not want to process a notarised tx if they receive it well outside the predefined TimeWindow, even if it has a valid _notary_timestamp_ (this is application specific).

**Comment by kchalkias** [comment]
As agreed with @mikehearn, I am closing (temporarily) this PR, until we have a concrete plan on timestamp handling/validation.

**Pull request closed without merging**